src/java.base/share/classes/java/util/concurrent/DelayScheduler.java

```diff
@@ -0,0 +1,568 @@
+        for (ScheduledForkJoinTask<?> a = (ScheduledForkJoinTask<?>)
+                 U.getAndSetReference(this, PENDING, null);
+             a != null; a = a.nextPending)
+            a.trySetCancelled(); // clear pending requests
+        return -1;
+    }
+
+    private static void cancelAll(ScheduledForkJoinTask<?>[] h, int n) {
+        if (h != null && h.length >= n) {
+            ScheduledForkJoinTask<?> t;
+            for (int i = 0; i < n; ++i) {
+                if ((t = h[i]) != null) {
+                    h[i] = null;
+                    t.heapIndex = -1;
+                    t.trySetCancelled();
+                }
+            }
+        }
+    }
+
+    /**
+     * Task class for DelayScheduler operations
+     */
+    @SuppressWarnings("serial")    // Not designed to be serializable
+    static final class ScheduledForkJoinTask<T>
+        extends ForkJoinTask.InterruptibleTask<T>
+        implements ScheduledFuture<T> {
+        ForkJoinPool pool;         // nulled out after use
+        Runnable runnable;         // at most one of runnable or callable nonnull
+        Callable<? extends T> callable;
+        T result;
+        ScheduledForkJoinTask<?> nextPending; // for DelayScheduler pending queue
+        long when;                  // nanoTime-based trigger time
+        final long nextDelay;       // 0: once; <0: fixedDelay; >0: fixedRate
+        int heapIndex;              // if non-negative, index on heap
+        final boolean isImmediate;  // run by scheduler vs submitted when ready
+
+        /**
+         * Creates a new ScheduledForkJoinTask
+         * @param delay initial delay, in nanoseconds
+         * @param nextDelay 0 for one-shot, negative for fixed delay,
+         *        positive for fixed rate, in nanoseconds
+         * @param isImmediate if action is to be performed
+         *        by scheduler versus submitting to a WorkQueue
+         * @param runnable action (null if implementing callable version)
+         * @param callable function (null if implementing runnable versions)
+         * @param pool the pool for resubmissions and cancellations
+         *        (disabled if null)
+         */
+        public ScheduledForkJoinTask(long delay, long nextDelay,
+                                     boolean isImmediate, Runnable runnable,
+                                     Callable<T> callable, ForkJoinPool pool) {
+            this.when = DelayScheduler.now() + Math.max(delay, 0L);
+            this.heapIndex = -1;
+            this.nextDelay = nextDelay;
+            this.isImmediate = isImmediate;
+            this.runnable = runnable;
+            this.callable = callable;
+            this.pool = pool;
+        }
+
+        public void schedule() { // relay to pool, to allow independent use
+            ForkJoinPool p;
+            if ((p = pool) != null) // else already run
+                p.scheduleDelayedTask(this);
+        }
+
+        // InterruptibleTask methods
```

Comment:
@DougLea I think a few methods below should have @Override annotations.

Reply:
Not sure of conventions for overrides of non-public methods. The internal FJT classes only  contain them, and I uniformly omit annotations. Do they add any value?